### PR TITLE
Add Unit test to verify Telemetry Output for acquireToken 

### LINKED ISSUE #832: Add telemetry validation to the acquireToken tests
The only testing of telemetry is testing the API surface itself, there is no validation of the telemetry stream coming out of library usage. We minimally should add telemetry validation to our acquireToken tests.

### PR DESCRIPTION
**Pull request by jasoncoolmax**
(fixes #832)

- Unit test is added to verify telemetry output for acquireToken.
- Remove some unnecessary check in unit tests
- ADTelemetry.h is made public

### CODE REVIEW

**Comment by jasoncoolmax** [comment]
Sure!

File: ADAL/src/telemetry/ADTelemetry.m [hunk 1]
```diff
@@ -67,6 +67,12 @@ - (void)registerDispatcher:(id<ADDispatcher>)dispatcher
     @synchronized(self)
     {
         SAFE_ARC_RELEASE(_dispatcher);
+        if (!dispatcher)
+        {
+            _dispatcher = nil;
+            return;
+        }
+
```
Comment (jasoncoolmax) [inline ADAL/src/telemetry/ADTelemetry.m]: if user passes in nil as dispatcher, we simply set _dispatcher to nil, such that we won't end up collecting a bunch of events and has no where to dispatch and discard them.

File: ADAL/src/request/ADAuthenticationRequest+AcquireToken.m [hunk 2]
```diff
@@ -80,9 +80,16 @@ - (void)acquireToken:(NSString *)eventName
         
         [event setCorrelationId:self.correlationId];
         [event setClientId:_requestParams.clientId];
+        [event setAuthority:_context.authority];
         [event setExtendedExpiresOnSetting:[_requestParams extendedLifetime]? AD_TELEMETRY_YES:AD_TELEMETRY_NO];
         [event setPromptBehavior:_promptBehavior];
-        [event setUserInformation:[[result tokenCacheItem] userInformation]];
+        if ([result tokenCacheItem])
+        {
+            [event setUserInformation:[[result tokenCacheItem] userInformation]];
```
Comment (RPangrle) [inline ADAL/src/request/ADAuthenticationRequest+AcquireToken.m]: Please use consistent dot syntax. As a rule-of-thumb if it's a property use dot syntax, so this should be `result.tokenCacheItem.UserInformation`

File: ADAL/src/request/ADAuthenticationRequest+AcquireToken.m [hunk 3]
```diff
@@ -80,9 +80,16 @@ - (void)acquireToken:(NSString *)eventName
...
+        }
+        else{
```
Comment (RPangrle) [inline ADAL/src/request/ADAuthenticationRequest+AcquireToken.m]: `{` on new line

File: ADAL/src/telemetry/ADTelemetryAPIEvent.h [hunk 4]
```diff
@@ -30,6 +30,7 @@
 - (void)setExtendedExpiresOnSetting:(NSString*)extendedExpiresOnSetting;
 - (void)setPromptBehavior:(ADPromptBehavior)promptBehavior;
 - (void)setUserInformation:(ADUserInformation*)userInfo;
+- (void)setUserId:(NSString*)userId;
```
Comment (RPangrle) [inline ADAL/src/telemetry/ADTelemetryAPIEvent.h]: There should be a space between `NSString` and `*`

File: ADAL/src/telemetry/ADTelemetryAPIEvent.m [hunk 5]
```diff
@@ -67,6 +67,11 @@ - (void)setUserInformation:(ADUserInformation*)userInfo
     [self setProperty:AD_TELEMETRY_IDP value:[userInfo identityProvider]];
 }
 
+- (void)setUserId:(NSString*)userId
```
Comment (RPangrle) [inline ADAL/src/telemetry/ADTelemetryAPIEvent.m]: Space between `NSString` and `*`

File: ADAL/tests/ADAcquireTokenTests.m [hunk 6]
```diff
@@ -586,6 +653,111 @@ - (void)testSilentExpiredATBadMRRT
      }];
     
     TEST_WAIT;
+    
+    // verify telemetry output
+    // there should be 2 telemetry events recorded as there are 2 acquire token calls
+    XCTAssertEqual([receivedEvents count], 2);
+    
+    // the following properties are expected for the 1st acquire token call
+    XCTAssertEqual([self adGetPropertyCount:[receivedEvents firstObject]
```
Comment (RPangrle) [inline ADAL/tests/ADAcquireTokenTests.m]: Pull out `[receivedEvents firstObject]` into a local variable and use that instead.

File: ADAL/tests/ADAcquireTokenTests.m [hunk 7]
```diff
@@ -986,6 +1167,62 @@ - (void)testAcquireTokenUsingFRT
      }];
     
     TEST_WAIT;
+    
+    // verify telemetry output
+    // there should be 1 telemetry events recorded as aggregation flag is ON
+    XCTAssertEqual([receivedEvents count], 1);
+    
+    // the following properties are expected in an aggregrated event
+    XCTAssertEqual([self adGetPropertyCount:[receivedEvents firstObject]
```
Comment (RPangrle) [inline ADAL/tests/ADAcquireTokenTests.m]: Pull out `[receivedEvents firstObject]` into a local variable and use that instead.

File: ADAL/tests/XCTestCase+TestHelperMethods.h [hunk 8]
```diff
@@ -210,6 +211,14 @@ typedef enum
                          line:(int)line
                     semaphore:(dispatch_semaphore_t)signal
                         block:(void (^)(void)) block;
+
+//The method retrieves property value from a telemetry event by property name
+- (NSString*)adGetPropertyFromEvent:(NSArray*)event
```
Comment (RPangrle) [inline ADAL/tests/XCTestCase+TestHelperMethods.h]: Space between `NSArray` and `*`

File: ADAL/tests/XCTestCase+TestHelperMethods.h [hunk 9]
```diff
@@ -210,6 +211,14 @@ typedef enum
...
+                     propertyName:(NSString*)propertyName;
```
Comment (RPangrle) [inline ADAL/tests/XCTestCase+TestHelperMethods.h]: Space between `NSString` and `*`

File: ADAL/tests/XCTestCase+TestHelperMethods.h [hunk 10]
```diff
@@ -210,6 +211,14 @@ typedef enum
...
+
+//The method counts the properties in a telemetry event matching propertyName
+- (NSInteger)adGetPropertyCount:(NSArray*)event
```
Comment (RPangrle) [inline ADAL/tests/XCTestCase+TestHelperMethods.h]: Space between `NSArray` and `*`

File: ADAL/tests/XCTestCase+TestHelperMethods.h [hunk 11]
```diff
@@ -210,6 +211,14 @@ typedef enum
...
+                 propertyName:(NSString*)propertyName;
```
Comment (RPangrle) [inline ADAL/tests/XCTestCase+TestHelperMethods.h]: Space between `NSString` and `*`

File: ADAL/tests/XCTestCase+TestHelperMethods.m [hunk 12]
```diff
@@ -508,4 +508,31 @@ - (ADTestURLResponse *)adResponseRefreshToken:(NSString *)oldRefreshToken
 
 }
 
+- (NSString*)adGetPropertyFromEvent:(NSArray*)event
```
Comment (RPangrle) [inline ADAL/tests/XCTestCase+TestHelperMethods.m]: Space between `NSArray` and `*`

File: ADAL/tests/XCTestCase+TestHelperMethods.m [hunk 13]
```diff
@@ -508,4 +508,31 @@ - (ADTestURLResponse *)adResponseRefreshToken:(NSString *)oldRefreshToken
...
+    {
+        if ([property.name isEqualToString:propertyName])
+        {
+            return property.value;
+        }
+    }
+    return nil;
+}
+
+- (NSInteger)adGetPropertyCount:(NSArray*)event
```
Comment (RPangrle) [inline ADAL/tests/XCTestCase+TestHelperMethods.m]: Space between `NSArray` and `*`

File: ADAL/tests/XCTestCase+TestHelperMethods.m [hunk 14]
```diff
@@ -508,4 +508,31 @@ - (ADTestURLResponse *)adResponseRefreshToken:(NSString *)oldRefreshToken
 
 }
 
+- (NSString*)adGetPropertyFromEvent:(NSArray*)event
+                     propertyName:(NSString*)propertyName
```
Comment (RPangrle) [inline ADAL/tests/XCTestCase+TestHelperMethods.m]: Space between `NSString` and `*`

File: ADAL/tests/XCTestCase+TestHelperMethods.m [hunk 15]
```diff
@@ -508,4 +508,31 @@ - (ADTestURLResponse *)adResponseRefreshToken:(NSString *)oldRefreshToken
...
+        if ([property.name isEqualToString:propertyName])
+        {
+            return property.value;
+        }
+    }
+    return nil;
+}
+
+- (NSInteger)adGetPropertyCount:(NSArray*)event
+                 propertyName:(NSString*)propertyName
```
Comment (RPangrle) [inline ADAL/tests/XCTestCase+TestHelperMethods.m]: Space between `NSString` and `*`

File: ADAL/src/request/ADAuthenticationRequest+AcquireToken.m (hunk 2 above)
Comment (jasoncoolmax) [inline ADAL/src/request/ADAuthenticationRequest+AcquireToken.m]: done

File: ADAL/src/request/ADAuthenticationRequest+AcquireToken.m (hunk 3 above)
Comment (jasoncoolmax) [inline ADAL/src/request/ADAuthenticationRequest+AcquireToken.m]: my bad

File: ADAL/src/telemetry/ADTelemetryAPIEvent.h (hunk 4 above)
Comment (jasoncoolmax) [inline ADAL/src/telemetry/ADTelemetryAPIEvent.h]: Sure, I will fix it :)

I notice there is an inconsistency for this issue in our code.
Probably we may need to go through the code in the future and make it consistent.

File: ADAL/src/telemetry/ADTelemetryAPIEvent.m (hunk 5 above)
Comment (jasoncoolmax) [inline ADAL/src/telemetry/ADTelemetryAPIEvent.m]: done

File: ADAL/tests/ADAcquireTokenTests.m (hunk 6 above)
Comment (jasoncoolmax) [inline ADAL/tests/ADAcquireTokenTests.m]: :+1:
It's not your first time to point that out, I will keep that in mind :)

File: ADAL/tests/ADAcquireTokenTests.m (hunk 7 above)
Comment (jasoncoolmax) [inline ADAL/tests/ADAcquireTokenTests.m]: done

File: ADAL/tests/XCTestCase+TestHelperMethods.h (hunk 8 above)
Comment (jasoncoolmax) [inline ADAL/tests/XCTestCase+TestHelperMethods.h]: done

File: ADAL/tests/XCTestCase+TestHelperMethods.h (hunk 9 above)
Comment (jasoncoolmax) [inline ADAL/tests/XCTestCase+TestHelperMethods.h]: done

File: ADAL/tests/XCTestCase+TestHelperMethods.h (hunk 10 above)
Comment (jasoncoolmax) [inline ADAL/tests/XCTestCase+TestHelperMethods.h]: done

File: ADAL/tests/XCTestCase+TestHelperMethods.h (hunk 11 above)
Comment (jasoncoolmax) [inline ADAL/tests/XCTestCase+TestHelperMethods.h]: done

File: ADAL/tests/XCTestCase+TestHelperMethods.m (hunk 12 above)
Comment (jasoncoolmax) [inline ADAL/tests/XCTestCase+TestHelperMethods.m]: done

File: ADAL/tests/XCTestCase+TestHelperMethods.m (hunk 14 above)
Comment (jasoncoolmax) [inline ADAL/tests/XCTestCase+TestHelperMethods.m]: done

File: ADAL/tests/XCTestCase+TestHelperMethods.m (hunk 13 above)
Comment (jasoncoolmax) [inline ADAL/tests/XCTestCase+TestHelperMethods.m]: done

File: ADAL/tests/XCTestCase+TestHelperMethods.m (hunk 15 above)
Comment (jasoncoolmax) [inline ADAL/tests/XCTestCase+TestHelperMethods.m]: done

File: ADAL/src/telemetry/ADTelemetryHttpEvent.m [hunk 16]
```diff
@@ -85,18 +85,7 @@ - (void)addAggregatedPropertiesToDictionary:(NSMutableDictionary*)eventToBeDispa
...
-        if ([property.name isEqualToString:AD_TELEMETRY_HTTP_RESPONSE_CODE]
-            ||[property.name isEqualToString:AD_TELEMETRY_OAUTH_ERROR_CODE]
-            ||[property.name isEqualToString:AD_TELEMETRY_HTTP_ERROR_DOMAIN]
-            ||[property.name isEqualToString:AD_TELEMETRY_HTTP_PATH]
-            ||[property.name isEqualToString:AD_TELEMETRY_HTTP_REQUEST_ID_HEADER])
-        {
-            [eventToBeDispatched setObject:property.value forKey:property.name];
-        }
-    }
+    //TODO: output http event properties to aggregated event
```
Comment (jasoncoolmax) [inline ADAL/src/telemetry/ADTelemetryHttpEvent.m]: How to output http properties to aggregated event remains undecided according to the design doc.

File: ADAL/src/telemetry/ADTelemetryAPIEvent.h (hunk 4 above)
Comment (RPangrle) [inline ADAL/src/telemetry/ADTelemetryAPIEvent.h]: Yeah. The comments I made are to match the style used in Apple's headers. I try to clean up files when I'm around them, but yes, there are still places that need to be cleaned.

**Pull request merged**